### DISCUSSION TITLE
Typeahead with CLOG?

### DISCUSSION THREAD

**Discussion by vindarel** [rabbibotton/clog#111] [Ideas]
Hello,

One important widget for my apps is a typeahead with Ajax calls. For example, [Twitter TypeAhead.js](https://github.com/twitter/typeahead.js/). It is extremely useful: wait for input (and define a minimal input length), request a DB asynchronously, show the results (with formatted HTML, unlike the simple UI of HTML5's `datalist`). 

But that requires writing the JS machinery, defining an API endpoint, etc. I wonder if CLOG could come in-between and allow to use this or to write this type of control with less to no JS and more Lisp. 

How easy would that be to use CLOG here? Can it hook into typeahead.js, would we need to write such a control in "pure CLOG" from scratch?

I read that [typeahead.js can use WebSockets](https://github.com/twitter/typeahead.js/issues/338), if that helps.

Thanks again.

edit: an async `datalist` would be great already.

**Comment by rabbibotton** [comment]
Once a page load boot.js from CLOG you really can do all these things easily and certainly faster than an ajax (or even an on demand ws) solution.

For the datalist:
Once you have your data ready on the lisp side you just create a string containing the code to create the datalist on the browser then use JS-EXECUTE and it is there. No need for json, yaml, etc. In fact I have not used json or yaml in 10 years for web programming.

For full functional lisp component:
For the typeahead you just use ATTACH-AS-CHILD to attach to the form input then add events for changes, etc. easy enough to launch a thread to time things etc. There really is no need for js.

What I would do:
Create a CLOG plugin from it. It is a jQuery plugin so even easier. In my haste to get out CLOG and builder I still have not done docs for plugins (which are even cooler now as easy to add them to CLOG Builder as well).

This would be excellent first plugin to do, so think I'll do it soon

**Comment by rabbibotton** [comment]
[clog-typeahead.zip](https://github.com/rabbibotton/clog/files/8083688/clog-typeahead.zip)

Here is a plugin for using typeahead.js  I will do a write up about it later.

The test in the lisp file shows how to use with a CLOG app. To use with some random page, you add boot.js, load the typeahead java scriptfile, when that file connects to the clog server you attach to the html ids you want to use and the rest is the same.

**Reply by vindarel** [reply]
Wow. Impressive. I'll try a bit later. Thanks!

**Reply by vindarel** [reply]
I run start-test and I get:

```
CL-USER> (clog-typeahead::start-test )
Hunchentoot server is started.
Listening on 0.0.0.0:8080.
HTTP listening on    : 0.0.0.0:8080
HTML Root            : /home/vince/bacasable/lisp-projects/clog-typeahead/./www/
Boot js source       : compiled in
Boot file for path / : /boot.html
NIL
NIL
0

=>

The value
  NIL
is not of type
  (OR FILE-STREAM SYNONYM-STREAM BROADCAST-STREAM)
when binding STREAM
   [Condition of type TYPE-ERROR]

Backtrace for: #<SB-THREAD:THREAD "hunchentoot-worker-127.0.0.1:54822" RUNNING {10081BF853}>
0: (FILE-LENGTH NIL) [external]
1: ((LAMBDA (CLOG-CONNECTION::ENV) :IN CLOG-CONNECTION:INITIALIZE) (:REQUEST-METHOD :GET :SCRIPT-NAME "" :PATH-INFO "/" :SERVER-NAME "127.0.0.1" :SERVER-PORT 8080 :SERVER-PROTOCOL :HTTP/1.1 ...))
2: ((LAMBDA (LACK.MIDDLEWARE.BACKTRACE::ENV) :IN "/home/vince/quicklisp/dists/quicklisp/software/lack-20210807-git/src/middleware/backtrace.lisp") (:REQUEST-METHOD :GET :SCRIPT-NAME "" :PATH-INFO "/" :SERVER-NAME "127.0.0.1" :SERVER-PORT 8080 :SERVER-PROTOCOL :HTTP/1.1 ...))
[…]
Request:
    REQUEST-METHOD: :GET
    SCRIPT-NAME: ""
    PATH-INFO: "/"
    SERVER-NAME: "127.0.0.1"
    SERVER-PORT: 8080
    SERVER-PROTOCOL: :HTTP/1.1
    REQUEST-URI: "/"
    URL-SCHEME: "http"
    REMOTE-ADDR: "127.0.0.1"
    REMOTE-PORT: 54822
    QUERY-STRING: NIL
    RAW-BODY: #<FLEXI-STREAMS:FLEXI-IO-STREAM {10090B3F03}>
    CONTENT-LENGTH: NIL
    CONTENT-TYPE: NIL
    CLACK.STREAMING: T
    CLACK.IO: #<CLACK.HANDLER.HUNCHENTOOT::CLIENT {1009186F13}>
    HEADERS:
        host: "127.0.0.1:8080"
        user-agent: "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:90.0) Gecko/20100101 Firefox/90.0"
        accept: "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8"
        accept-language: "en-GB,en;q=0.5"
        accept-encoding: "gzip, deflate"
        dnt: "1"
        connection: "keep-alive"
        cookie: "_hellophoenix_key=SFMyNTY.g3QAAAABbQAAAAtfY3NyZl90b2tlbm0AAAAYWkFjdW9idVJRTEtIV0kwQWo2NElSOUZr.I23WIYZ6OtCJ1I0U3Ubc2lLJKIlTSB7ZGaGhtUEdBn4"
        upgrade-insecure-requests: "1"
        sec-fetch-dest: "document"
        sec-fetch-mode: "navigate"
        sec-fetch-site: "cross-site"
```

I downloaded the file typeahead.jquery.js into `www/js/`.

Do I have to do something else?

**Reply by rabbibotton** [reply]
Not sure I understand, make sure using latest CLOG. Unzip in to you lisp project dir, the file clog-typeahead.zip. Start slime, (ql:quickload :clog-typeahead) (clog-typeahead:start-test) - if you previously initialized CLOG you should (clog:shutdown) first.

**Reply by vindarel** [reply]
I confess I was missing a few commits of CLOG, but same error today. I restarted my Lisp process, loaded clog.asd and clog-typeahead.asd, quickloaded clog-typeahead, started `(clog-typeahead::start-test )`. SBCL 1.4.5.

It happens here, in clog-connection.lisp:

```lisp
		      (let ((file (uiop:subpathname static-root clog-path)))
			(with-open-file (stream file :direction :input
						     :if-does-not-exist nil)
			  (let ((page-data (make-string (file-length stream)))
```

(edit) "file" here has the value `/home/vince/bacasable/lisp-projects/clog-typeahead/./www/boot.html`.

In your zip archive, there is nothing inside the `www/` directory, is that normal? I created `www/js/` and downloaded `typeahead.jquery.js` there.

**Reply by vindarel** [reply]
ah. I didn't create any book.[js, html] file myself. Maybe I should.

**Reply by rabbibotton** [reply]
zip  is apparently messed. Here is correct one

[clog-typeahead.zip](https://github.com/rabbibotton/clog/files/8112621/clog-typeahead.zip)

**Reply by vindarel** [reply]
Now I see something!! thanks.

![Selection_054](https://user-images.githubusercontent.com/3721004/155181883-9bbf4e01-d72d-417c-8eb9-35ca14b972b5.png)

**Comment by vindarel** [comment]
So, thanks again. Now what I need to investigate is:

- how to catch the event of the user input, and return a list of candidates (the ajax stuff) (currently the plugin is attached with a given list of items),
- how to make the plugin look as in the demo

We got this:

![Selection_054](https://user-images.githubusercontent.com/3721004/155196275-4d7db4b9-dd9d-4a15-8679-c5a9a054284f.png)

instead of:

![Selection_055](https://user-images.githubusercontent.com/3721004/155196378-1dce12fb-546c-452c-bf2b-8bc5e6ce7864.png)

**Comment by rabbibotton** [comment]
I added call backs from javascript to clog as events that can be set on the CLOG side with set-on-event or set-on-event-with-data and called from the browser side using jQueries .trigger("costum-event-name", "data if desired") 

This means this could be expanded on with more flexibility now.

**Comment by rabbibotton** [comment]
I have updated https://docs.google.com/document/d/1EDODcnyijP_EjrDQTiA0AB-zbMQzvCo2HltXarw5q5A/edit?usp=sharing

The example now contains a callback version of typeahead.js so now you can for example based on what a user types return a database query easily etc.

```
(attach-typeahead element2 (lambda (obj data)
				 (declare (ignore obj data))
					; ideally we should use DATA to determine
					; what list is returned
				 '("one" "two" "three" "four")))))
```
data contains whatever the user types and the lookahead is whatever list you return.

**Reply by rabbibotton** [reply]
I added a repo with it to - https://github.com/rabbibotton/clog-typeahead

Going to now work on needed API for CLOG Builder plug-ins and use this as first test.

**Comment by rabbibotton** [comment]
I update the plugin doc (will make another update when builder stuff perfect) - https://docs.google.com/document/d/1EDODcnyijP_EjrDQTiA0AB-zbMQzvCo2HltXarw5q5A

and also the repo now has a full plugin that works with builder (still working on improving)
one you git clone  https://github.com/rabbibotton/clog-typeahead in to ~/common-lisp
you do (to try docs on using to come):

(ql:quickload :clog-typeahead/tools)
(clog-tools:clog-builder :system :clog-typeahead)

**Comment by rabbibotton** [comment]
Done :)

The github is at:

https://github.com/rabbibotton/clog-typeahead



The tutorial on creating a custom plug-in is at:

https://docs.google.com/document/d/1EDODcnyijP_EjrDQTiA0AB-zbMQzvCo2HltXarw5q5A/

**Comment by vindarel** [comment]
This example confirms it is possible to have a decent ajax autocomplete without using a JS library: https://github.com/rajasegar/cl-beers/blob/main/templates/layouts/default.html#L39

```html
<form class="d-flex">
        <input id="txtSearch" class="form-control me-2" type="search" name="query"
               placeholder="Search Beer..." aria-label="Search Beer"
               hx-post="/search"
               hx-trigger="keyup changed delay:500ms"
               hx-target="#search-results"
               _="on htmx:afterOnLoad add .show to #search-results">
        <ul id="search-results" class="dropdown-menu mt-5">
        </ul>
      </form>
```

It uses Bootstrap's dropdown element, the goal is to add `<ul>` elements with a "dropdown-item" class, and to toggle the visibility of the menu list. Usually we can display any HTML in the element. 

I am trying with Bulma CSS.

Still, advanced libraries can highlight the matching text in elements and other niceties.

**Reply by rabbibotton** [reply]
Not sure I understand as that example heavily relies on Ajax and JavaScript and in my opinion in the worst way by hiding behind modifications to html itself in non compatible ways.

I didn't realize you were still looking for a solution, this is super trivial in CLOG especially with the builder, so not sure why spending time on it.

Boostrap (specifically for this exact case) also does something similar by lacing the HTML with code and creating some minor issue with using it with the builder.

**Reply by vindarel** [reply]
Sure this example relies on Ajax, but not on a full JS library. Thus it should be trivial to replace the Ajax calls with CLOG.

I also find it nice as it shows the basic UI element we need is a dropdown to present the results.

**Reply by rabbibotton** [reply]
got it, let me know if one exactly fits your needs.
